edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs

```diff
@@ -77,72 +89,149 @@ public static (IList<X509Certificate2>, Option<string>) BuildCertificateList(X50
+            return ValidateClientCert(remoteCertificate, remoteCertificateChain, trustedCACerts, logger);
+        }
 
-            if (!caChainCerts.HasValue)
+        public static bool IsCACertificate(X509Certificate2 certificate)
+        {
+            // https://tools.ietf.org/html/rfc3280#section-4.2.1.3
+            // The keyCertSign bit is asserted when the subject public key is
+            // used for verifying a signature on public key certificates.  If the
+            // keyCertSign bit is asserted, then the cA bit in the basic
+            // constraints extension (section 4.2.1.10) MUST also be asserted.
+
+            // https://tools.ietf.org/html/rfc3280#section-4.2.1.10
+            // The cA boolean indicates whether the certified public key belongs to
+            // a CA.  If the cA boolean is not asserted, then the keyCertSign bit in
+            // the key usage extension MUST NOT be asserted.
+            var basicConstraints = certificate.Extensions.OfType<X509BasicConstraintsExtension>();
+            foreach (X509BasicConstraintsExtension extension in basicConstraints)
             {
-                logger.LogWarning($"Cannot validate certificate with subject: {certificate.Subject}. Hub CA cert chain missing.");
-                return false;
+                if (extension.CertificateAuthority)
+                {
+                    return true;
+                }
             }
 
+            return false;
+        }
+
+        public static bool ValidateClientCert(X509Certificate2 certificate, IList<X509Certificate2> certificateChain, Option<IList<X509Certificate2>> trustedCACerts, ILogger logger)
+        {
+            Preconditions.CheckNotNull(certificate);
+            Preconditions.CheckNotNull(certificateChain);
+            Preconditions.CheckNotNull(trustedCACerts);
+            Preconditions.CheckNotNull(logger);
+
             DateTime currentTime = DateTime.Now;
-            if (newCert.NotAfter < currentTime)
+
+            if (certificate.NotAfter < currentTime)
             {
-                logger.LogWarning($"Certificate with subject: {newCert.Subject} has expired");
+                logger.LogWarning($"Certificate with subject: {certificate.Subject} has expired");
                 return false;
             }
 
-            if (newCert.NotBefore > currentTime)
+            if (certificate.NotBefore > currentTime)
             {
-                logger.LogWarning($"Certificate with subject: {newCert.Subject} is not valid");
+                logger.LogWarning($"Certificate with subject: {certificate.Subject} is not valid");
                 return false;
             }
 
-            IList<X509Certificate2> certChain = GetCertificatesFromChain(chain);
+            if (IsCACertificate(certificate))
+            {
+                logger.LogWarning($"Client certificate with basic constraints as TRUE is not permitted");
```

Comment:
Should we make this log a bit more user friendly?

Reply:
Thanks, fixed.